### PR TITLE
ZHA UI enhancements

### PR DESCRIPTION
**Pull request by dmulcahey**
## Proposed change

This PR adds auto zoom in the ZHA network visualization page and it also adds a dialog on the device page that allows you to see the children for router and coordinator type devices. 

auto-zoom

Screen Shot 2021-03-06 at 4 59 19 PM

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

File: src/panels/config/integrations/integration-panels/zha/dialog-zha-device-children.ts [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+      const devices = await fetchDevices(this.hass!);
+      this._devices = new Map(
+        devices.map((device: ZHADevice) => [device.ieee, device])
+      );
+    }
+  }
+
+  private _close(): void {
+    this._device = undefined;
+  }
```
Comment (bramkragten) [inline src/panels/config/integrations/integration-panels/zha/dialog-zha-device-children.ts:119]:
```suggestion
```

File: src/panels/config/integrations/integration-panels/zha/dialog-zha-device-children.ts [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    },
+  };
+
+  public async showDialog(
+    params: ZHADeviceChildrenDialogParams
+  ): Promise<void> {
+    this._device = params.device;
+    await this._fetchData();
+  }
+
```
Comment (bramkragten) [inline src/panels/config/integrations/integration-panels/zha/dialog-zha-device-children.ts:80]:
```suggestion
  public closeDialog(): void {
    this._device = undefined;
    fireEvent(this, "dialog-closed", { dialog: this.localName });
  }
```

File: src/panels/config/integrations/integration-panels/zha/dialog-zha-device-children.ts [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+  protected render(): TemplateResult {
+    if (!this._device) {
+      return html``;
+    }
+
+    return html`
+      <ha-dialog
+        open
+        hideActions
+        @closing="${this._close}"
```
Comment (bramkragten) [inline src/panels/config/integrations/integration-panels/zha/dialog-zha-device-children.ts:90]:
```suggestion
        @closed="${this.closeDialog}"
```

File: src/panels/config/integrations/integration-panels/zha/dialog-zha-device-children.ts [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+      direction: "asc",
+      width: "75px",
+    },
+  };
+
+  public async showDialog(
+    params: ZHADeviceChildrenDialogParams
+  ): Promise<void> {
+    this._device = params.device;
+    await this._fetchData();
```
Comment (bramkragten) [inline src/panels/config/integrations/integration-panels/zha/dialog-zha-device-children.ts:78]: Don't await this, and show a loader in the dialog until `this._devices` is loaded.

**Review by bramkragten — APPROVED** [review]

File: src/panels/config/integrations/integration-panels/zha/dialog-zha-device-children.ts (hunk 4 above)
Comment (dmulcahey) [reply src/panels/config/integrations/integration-panels/zha/dialog-zha-device-children.ts:78]: Is there another example of this that I can look at?

Comment (bramkragten) [reply src/panels/config/integrations/integration-panels/zha/dialog-zha-device-children.ts:78]: Just remove that await and add this in render:
```ts
${!this._devices ? html`<ha-circular-progress alt="Loading" size="large" active>
              </ha-circular-progress>` : html`<ha-data-table></ha-data-table>`}
              ```

**Comment by dmulcahey** [comment]
done

File: src/panels/config/integrations/integration-panels/zha/dialog-zha-device-children.ts [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+    this._device = undefined;
+    this._devices = undefined;
+    this._opened = false;
+    fireEvent(this, "dialog-closed", { dialog: this.localName });
+  }
+
+  protected render(): TemplateResult {
+    return html`
+      <ha-dialog
+        .open=${this._opened}
```
Comment (bramkragten) [inline src/panels/config/integrations/integration-panels/zha/dialog-zha-device-children.ts:99]: I would remove this and add back your old check if _device is undefined and render blank.

Comment (dmulcahey) [reply src/panels/config/integrations/integration-panels/zha/dialog-zha-device-children.ts:99]: The .open part? Without it the dialog only opens the 1st time it’s launched. All subsequent launches show nothing...

Comment (dmulcahey) [reply src/panels/config/integrations/integration-panels/zha/dialog-zha-device-children.ts:99]: Or are you saying the spinner shouldn’t be in the dialog?

Comment (bramkragten) [reply src/panels/config/integrations/integration-panels/zha/dialog-zha-device-children.ts:99]: Change the closing event to closed

Comment (dmulcahey) [reply src/panels/config/integrations/integration-panels/zha/dialog-zha-device-children.ts:99]: I think I tried that. I’ll try again in a bit.

Comment (dmulcahey) [reply src/panels/config/integrations/integration-panels/zha/dialog-zha-device-children.ts]: updated it again.... I think this is what you meant

File: src/panels/config/integrations/integration-panels/zha/dialog-zha-device-children.ts [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+
+  protected render(): TemplateResult {
+    if (!this._device) {
+      return html``;
+    }
+    return html`
+      <ha-dialog
+        hideActions
+        open
+        @closed="${this.closeDialog}"
```
Comment (bramkragten) [inline src/panels/config/integrations/integration-panels/zha/dialog-zha-device-children.ts:100]:
```suggestion
        @closed=${this.closeDialog}
```

File: src/panels/config/integrations/integration-panels/zha/dialog-zha-device-children.ts [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+    lqi: {
+      title: "LQI",
+      sortable: true,
+      filterable: true,
+      direction: "asc",
+      width: "75px",
+    },
+  };
+
+  public async showDialog(
```
Comment (bramkragten) [inline src/panels/config/integrations/integration-panels/zha/dialog-zha-device-children.ts:79]:
```suggestion
  public showDialog(
```

**Review by bramkragten — APPROVED** [review]

File: src/panels/config/integrations/integration-panels/zha/dialog-zha-device-children.ts [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+      sortable: true,
+      filterable: true,
+      direction: "asc",
+      width: "75px",
+    },
+  };
+
+  public showDialog(
+    params: ZHADeviceChildrenDialogParams
+  ): Promise<void> {
```
Comment (bramkragten) [inline src/panels/config/integrations/integration-panels/zha/dialog-zha-device-children.ts:81]:
```suggestion
  ): void {
```

**Pull request merged**